woocommerce/woocommerce-ios
# [Woo POS] Update alert webview to be presented modally instead of replacing the alert

Amends https://github.com/woocommerce/woocommerce-ios/pull/12932
Part of #12864 


## Why

As brought up in https://github.com/woocommerce/woocommerce-ios/pull/12932#discussion_r1625773595, the implementation to emit a card payment event equivalent to `showAlert` results in an issue where the webview can't be easily presented on top of the alert. This PR aims to fix this behavior, and allows web view presentation modally when an alert initiates a WC settings web view presentation.

## How

Based on @joshheald's helpful suggestions & feedback, this is a hopefully simple solution:

- A new `CardPresentPaymentsModalViewModelWCSettingsWebViewPresenting` for any alert view model to conform to, in order to provide a publisher of an optional web view model
  - `CardPresentModalConnectingFailedUpdateAddress`, the only alert that can present a WC Settings webview right now, conforms to this protocol and provides the web view model when the user taps on the CTA. The URL is DI'ed from the alert provider
- In order to show the web view as a sheet on top of the alert, while encapsulating the logic within the alert view, a new view model `CardPresentPaymentAlertSwiftUIViewModel` was created. It takes in an alert view model, and observes its web view model if it conforms to `CardPresentPaymentsModalViewModelWCSettingsWebViewPresenting`. This is the simplest way I could find while working with SwiftUI binding and the existing alert setup. If we plan to support other custom navigation initiated from the alert actions, we could follow a similar approach so that each alert can support all types of custom navigation

I didn't revert the UIKit implementation using `presentWCSettingsWebView` to add back the technical debt of having UI presentation logic mixed with the connection business logic, but I'm happy to do so in a separate PR.

The web view model was also renamed to "WC Settings" to keep the changes small. When we are ready to support other types of webview like showing receipt (sounds like we might not do this from the alert), we can update the protocol to make the web view more generic.

## Steps to reproduce


## Testing information


For easier testing, feel free to mock the error by replacing `CardReaderConnectionController.onInitialization` with

```swift
        let url = URL(string: "{{site_url}}/wp-admin/admin.php?page=wc-settings&tab=tax") // just an example
        showConnectionFailed(error: CardReaderServiceError.connection(underlyingError: .incompleteStoreAddress(adminUrl: url)))
```

And then:

Prerequisite: having a store eligible for POS

- Switch to a store in the prerequisite if needed
- Turn on the feature switch in Menu > Settings > Experimental Features > POS, if needed
- Go to the Menu tab --> the POS row should appear shortly
- Tap to enter POS
- Tap `Reader disconnected` --> the error modal should be shown
- Tap `Enter Address` --> it should open the webview modally (make sure the proxy is disabled)
- Drag to dismiss the webview --> it should go back to the alert without triggering retry
- Tap `Enter Address` again --> it should open the webview modally
- Tap `Done` --> it should retry the connection flow

## Screenshots


Sometimes the card reader scanning from the retry took a while during my testing, I'm not sure if it's from the replacement of the `CardReaderConnectionController.onInitialization` or some pre-existing issue. As we're focusing on the happy path, I made a note to test this case when we're handling all errors properly.


https://github.com/woocommerce/woocommerce-ios/assets/1945542/a5556f8f-041c-4b2f-acd1-1c81c722787e



---
